Fix #7498: Apply non-render blocking attributes to self-hosted Google Fonts

Miraeld opened the pull request:
Fixes #7498

# Description

Fixes #7499 
- Ensures self-hosted fonts behave consistently with external optimized fonts

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue).

## Detailed scenario

### What was tested

The template mentioned by QA.

### How to test

refer to issue

## Technical description

### Documentation

- Add is_preload_enabled() method to check auto_preload_fonts setting
- Apply media='print' and onload='this.media="all"' when preload is enabled
- Include noscript fallback for users without JavaScript
- Ensures self-hosted fonts behave consistently with external optimized fonts

# Mandatory Checklist

## Code validation

- [x] I validated all the Acceptance Criteria. If possible, provide screenshots or videos.
- [x] I triggered all changed lines of code at least once without new errors/warnings/notices.
- [ ] I implemented built-in tests to cover the new/changed code.

## Code style

- [x] I wrote a self-explanatory code about what it does.
- [x] I protected entry points against unexpected inputs.
- [x] I did not introduce unnecessary complexity.
- [x] Output messages (errors, notices, logs) are explicit enough for users to understand the issue and are actionnable.

## Unticked items justification
E2E tests for this.

jeawhanlee commented:
@Miraeld Based on the discussion in [this](https://group-onecom.slack.com/archives/C08EFCYRQ2X/p1763381674719889) thread, I think simply removing the hook: `rocket_disable_google_fonts_preload` and the callback should be enough. Tell me what you think.

Miraeld commented:
@copilot Conversation summary
We discovered that Google Fonts preload is currently disabled when using the "Self-Host Google Fonts" feature, and we are investigating whether this should be re-enabled.

Key Findings
Root Cause: The preload was disabled in PR #7105 (commit 9cf845244ffbca0635ec29d01d8db320bf4fa36b) to prevent duplicate link tags.

Original Problem: There was a duplication issue where the same <link /> tags were appearing multiple times in the HTML (not just duplicate attributes, but entire duplicate tags).

Current State:

We tested and found no duplication occurs with current code
It's been confirmed there's no duplication on the page now
The new "head system" (introduced after PR #7105) already prevents duplications
Why It's Fixed Now: When the new head system was introduced, the features were refactored to use it, which inherently prevents duplicate tags from being printed.

Proposed Solution
Remove the hook that disables preload (rocket_disable_google_fonts_preload and its callback) since:

The duplication issue no longer exists
Current implementation doesn't match acceptance criteria
Self-hosted fonts should be preloaded for performance (they're render-blocking)

And based on previous comment, We think simply removing the hook: `rocket_disable_google_fonts_preload` and the callback should be enough. 
Could you make it happen ?

Miraeld committed:
Fix #7498: Re-enable preload for self-hosted Google Fonts

- Remove rocket_disable_google_fonts_preload hook and callback methods
- Update AbstractGFOptimization to check auto_preload_fonts setting directly
- This restores preload functionality that was disabled in PR #7105
- The original duplication issue is now prevented by the new head system

The hook was originally added to prevent duplicate link tags, but the new
head system already handles this. Removing the hook allows self-hosted fonts
to be preloaded correctly when the 'Preload fonts' setting is enabled.

Fixes #7498

Miraeld committed:
Fix default value

Miraeld committed:
Revert filter

Miraeld committed:
Remove change on comment

jeawhanlee committed:
Reverted changes on docblock

Mai-Saad commented:
@Miraeld Thanks for the PR. While test execution, seems we have the following regression
- When RUCSS is applied, the self hosted GFs arenot preloaded
Steps:
1- fresh install and activation of PR
2- enable self hosted GF and RUCSS
3- visit URL
4- wait till used CSS is generated and revisit the URL => Google fonts arenot preloaded while it is preloaded on trunk + noscript is there

on PR 
<img width="1920" height="1080" alt="Screenshot from 2025-11-19 08-45-45" src="https://github.com/user-attachments/assets/461313d9-dafb-4728-8f2b-c3a42c76e35f" />

on Trunk
<img width="1920" height="1080" alt="Screenshot from 2025-11-19 08-45-53" src="https://github.com/user-attachments/assets/e8352828-39f3-46ce-9699-23a6b59b877c" />

Can you please check? /cc @DahmaniAdame

jeawhanlee commented:
@Mai-Saad This could be from this [issue](https://github.com/wp-media/wp-rocket/issues/7544) where we disabled RUCSS preload fonts. Can you check if this is the same behavior on develop since the related PR is already merged.

wordpressfan commented:
I think this is related to this change:
https://github.com/wp-media/wp-rocket/pull/7481/files#diff-c0b914d33ca62bbda2554bbc234f755eb6063d101234ac0ff2110875e321ddb3R387-R392

I'll take this PR over, I have a thing to validate then will push the fix.

Mai-Saad commented:
@jeawhanlee  You are right, no preload Google fonts on dev while RUCSS is on and host GF is on (the noscript is not there as on trunk)
<img width="1920" height="1080" alt="Screenshot from 2025-11-19 11-24-01" src="https://github.com/user-attachments/assets/d22f1f5c-6112-4476-947f-36f0e21d2cf1" />

@DahmaniAdame Can you please confirm expectations here :pray:

DahmaniAdame commented:
> Can you please confirm expectations here 🙏

RUCSS shouldn't preload fonts since https://github.com/wp-media/wp-rocket/issues/7544

wordpressfan committed:
add data attribute to remove this noscript tag when self host google fonts feature is enabled.

wordpressfan committed:
fix some tests

Mai-Saad commented:
working as expected 
[testrail-report-789.pdf](https://github.com/user-attachments/files/23630970/testrail-report-789.pdf)
